src/parser/syntax.rs

```diff
@@ -0,0 +1,745 @@
+        let b: F = (*b as u64).into();
+        if bs.peek().is_none() {
+            res.add_assign(b)
+        } else {
+            res.add_assign(b);
+            res.mul_assign(F::from(256u64));
+        }
+    }
+    res
+}
+
+pub fn parse_num_inner<F: LurkField>(
+    base: base::LitBase,
+) -> impl Fn(Span<'_>) -> IResult<Span<'_>, Num<F>, ParseError<Span<'_>, F>> {
+    move |from: Span<'_>| {
+        let (upto, bytes): (Span<'_>, Vec<u8>) = base::parse_litbase_le_bytes(base)(from)?;
+        let max_bytes = (F::ZERO - F::ONE).to_bytes();
+        let max_uint = num_bigint::BigUint::from_bytes_le(&max_bytes);
+        if num_bigint::BigUint::from_bytes_le(&bytes) > max_uint {
+            ParseError::throw(
+                from,
+                ParseErrorKind::NumLiteralTooBig(F::most_positive(), max_uint),
+            )
+        } else {
+            let f = f_from_le_bytes::<F>(&bytes);
+            if let Some(x) = f.to_u64() {
+                Ok((upto, Num::U64(x)))
+            } else {
+                Ok((upto, Num::Scalar(f)))
+            }
+        }
+    }
+}
+
+pub fn parse_num<F: LurkField>(
+) -> impl Fn(Span<'_>) -> IResult<Span<'_>, Syntax<F>, ParseError<Span<'_>, F>> {
+    move |from: Span<'_>| {
+        let (i, neg) = opt(tag("-"))(from)?;
+        let (i, base) = alt((
+            preceded(tag("0"), base::parse_litbase_code()),
+            success(base::LitBase::Dec),
+        ))(i)?;
+        let (i, num) = parse_num_inner(base)(i)?;
+        let (upto, denom) = opt(preceded(tag("/"), parse_num_inner(base)))(i)?;
+        let pos = Pos::from_upto(from, upto);
+        let mut tmp = Num::<F>::U64(0);
+        if neg.is_some() {
+            tmp -= num;
+        } else {
+            tmp = num;
+        }
+        if let Some(denom) = denom {
+            tmp /= denom;
+        }
+        Ok((upto, Syntax::Num(pos, tmp)))
+    }
+}
+
+pub fn parse_string<F: LurkField>(
+) -> impl Fn(Span<'_>) -> IResult<Span<'_>, Syntax<F>, ParseError<Span<'_>, F>> {
+    move |from: Span<'_>| {
+        let (upto, s) = string::parse_string('"')(from)?;
+        let pos = Pos::from_upto(from, upto);
+        Ok((upto, Syntax::String(pos, s)))
+    }
+}
+
+// hash syntax for chars
+pub fn parse_hash_char<F: LurkField>(
+) -> impl Fn(Span<'_>) -> IResult<Span<'_>, Syntax<F>, ParseError<Span<'_>, F>> {
+    |from: Span<'_>| {
+        let (i, _) = tag("#\\")(from)?;
+        let (upto, c) = alt((string::parse_unicode(), anychar))(i)?;
+        let pos = Pos::from_upto(from, upto);
+        Ok((upto, Syntax::Char(pos, c)))
+    }
+}
+
+pub fn parse_char<F: LurkField>(
+) -> impl Fn(Span<'_>) -> IResult<Span<'_>, Syntax<F>, ParseError<Span<'_>, F>> {
+    move |from: Span<'_>| {
+        let (i, _) = tag("'")(from)?;
+        let (i, s) = string::parse_string_inner1('\'', true, "()'")(i)?;
+        let (upto, _) = tag("'")(i)?;
+        let mut chars: Vec<char> = s.chars().collect();
+        if chars.len() == 1 {
```

Comment:
How would this behave if the quoted symbol is represented using two meta-characters, such as the tab character (`’\t’`)?

Reply:
it works fine, escape characters are handled by by `parse_string_inner1`, which calls

```
/// Parse an escaped character: \n, \t, \r, \u{00AC}, etc.
pub fn parse_escaped_char<'a, F: LurkField>(
    delim: char,
    must_escape: &'static str,
) -> impl Fn(Span<'a>) -> IResult<Span<'a>, char, ParseError<Span<'a>, F>> {
    move |from: Span<'a>| {
        let (i, _) = char('\\')(from)?;
        let (i, esc) = alt((
            parse_unicode(),
            value('\n', char('n')),
            value('\r', char('r')),
            value('\t', char('t')),
            value('\u{08}', char('b')),
            value('\u{0C}', char('f')),
            value('\\', char('\\')),
            value('/', char('/')),
            value('"', char('"')),
            value('\'', char('\'')),
            value(delim, char(delim)),
            one_of(must_escape),
        ))(i)?;
        Ok((i, esc))
    }
}
```

To generate a single `StringFragment::EscapedChar(char)` that then turns into a singe `char` in the string. 

I'll add

```
        assert!(test(parse_char(), "'\\t'", Some(char!('\t'))));
```

to `unit_parse_char` to demonstrate that codepath